x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/CrossClusterApiKeyRoleDescriptorBuilder.java

```diff
@@ -0,0 +1,128 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.action.apikey;
+
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.util.CollectionUtils;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+
+import java.util.Arrays;
+import java.util.List;
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+public class CrossClusterApiKeyRoleDescriptorBuilder {
+
+    private static final String[] CCS_CLUSTER_PRIVILEGE_NAMES = { "cross_cluster_search" };
+    private static final String[] CCR_CLUSTER_PRIVILEGE_NAMES = { "cross_cluster_replication" };
+    private static final String[] CCS_AND_CCR_CLUSTER_PRIVILEGE_NAMES = { "cross_cluster_search", "cross_cluster_replication" };
+    private static final String[] CCS_INDICES_PRIVILEGE_NAMES = { "read", "read_cross_cluster", "view_index_metadata" };
+    private static final String[] CCR_INDICES_PRIVILEGE_NAMES = { "cross_cluster_replication", "cross_cluster_replication_internal" };
+    private static final String ROLE_DESCRIPTOR_NAME = "cross_cluster";
+
+    @SuppressWarnings("unchecked")
+    public static final ConstructingObjectParser<CrossClusterApiKeyRoleDescriptorBuilder, Void> PARSER = new ConstructingObjectParser<>(
+        "cross_cluster_api_key_request_access",
+        false,
+        (args, v) -> new CrossClusterApiKeyRoleDescriptorBuilder(
+            (List<RoleDescriptor.IndicesPrivileges>) args[0],
+            (List<RoleDescriptor.IndicesPrivileges>) args[1]
+        )
+    );
+
+    static {
+        PARSER.declareObjectArray(
+            optionalConstructorArg(),
+            (p, c) -> RoleDescriptor.parseIndexWithPrivileges(ROLE_DESCRIPTOR_NAME, CCS_INDICES_PRIVILEGE_NAMES, p),
+            new ParseField("search")
+        );
+        PARSER.declareObjectArray(
+            optionalConstructorArg(),
+            (p, c) -> RoleDescriptor.parseIndexWithPrivileges(ROLE_DESCRIPTOR_NAME, CCR_INDICES_PRIVILEGE_NAMES, p),
+            new ParseField("replication")
+        );
+    }
+
+    private final List<RoleDescriptor.IndicesPrivileges> search;
+    private final List<RoleDescriptor.IndicesPrivileges> replication;
+
+    private CrossClusterApiKeyRoleDescriptorBuilder(
+        List<RoleDescriptor.IndicesPrivileges> search,
+        List<RoleDescriptor.IndicesPrivileges> replication
+    ) {
+        this.search = search == null ? List.of() : search;
+        this.replication = replication == null ? List.of() : replication;
+    }
+
+    public RoleDescriptor build() {
+        final String[] clusterPrivileges;
+        if (search.isEmpty() && replication.isEmpty()) {
+            throw new IllegalArgumentException("must specify non-empty access for either [search] or [replication]");
```

Comment:
Don't think it matters too much, but `ElasticsearchParseException` would align better with the RoleDescriptor parsing code.

Reply:
I'd prefer to keep it as is mainly because strictly speaking it is not a parsing exception. In both case, the error code will be 400 so it seems OK either way. Please let me know if you feel strongly.

Reply:
Sounds good!